File: 10/10.1.py
def solve(size, lengths):
  field = range(size)
  pos = 0
  skip = 0

  for le in lengths:
    p1 = pos
    p2 = ( pos + le - 1 ) % size

    print 'pos', pos
    print 'len', le
    print '[',p1,':',p2,']:',p1 - p2

    while True:
      if p1 == p2:
        break
      if p1 == p2 - 1:
        break
      if p2 == p1 - 1:
        break
      c = field[p2]
      field[p2] = field[p1]
      field[p1] = c
      p1 = ( p1 + 1 ) % size
      if p1 == p2:
        break
      p2 = ( p2 - 1) % size

    print field
    pos = ( pos + le + skip ) % size
    skip += 1

  return field[0] * field[1]

assert solve(5, [3, 4, 1, 5]) == 12;
print solve(256, [120,93,0,90,5,80,129,74,1,165,204,255,254,2,50,113]);

File: 10/10-1.js
function assert(a, b) {
  if (JSON.stringify(a) != JSON.stringify(b)) {
    console.log("FAIL", a, "!=", b);
    return;
  }
  console.log("PASS");
}

function solve(cmd, size) {
  const list = Array(size)
    .fill(0)
    .map((a, i) => i);

  let pos = 0;
  let skip = 0;
  while (cmd.length) {
    let len = cmd.shift();
    let end = pos + len - 1;
    let endNext = pos + len;

    while (pos < end) {
      const v1 = list[pos % list.length];
      const v2 = list[end % list.length];
      list[pos % list.length] = v2;
      list[end % list.length] = v1;
      pos++;
      end--;
    }
    pos = (endNext + skip) % list.length;
    skip++;
  }
  return list;
}

assert(solve([3, 4, 1, 5], 5), [3, 4, 2, 1, 0]);

const data = [
  120,
  93,
  0,
  90,
  5,
  80,
  129,
  74,
  1,
  165,
  204,
  255,
  254,
  2,
  50,
  113
];

const res = solve(data, 256);
console.log(res[0] * res[1]);
